src/daemon/distribution_test.c

```diff
@@ -0,0 +1,448 @@
+/**
+ * collectd - src/daemon/distribution_test.c
+ * Copyright (C) 2020       Google LLC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ *
+ * Authors:
+ *   Svetlana Shmidt <sshmidt at google.com>
+ */
+
+#include "collectd.h"
+
+#include "distribution.h"
+#include "testing.h"
+
+static double *linear_upper_bounds(size_t num, double size) {
+  double *linear_upper_bounds = calloc(num, sizeof(*linear_upper_bounds));
+  for (size_t i = 0; i + 1 < num; i++)
+    linear_upper_bounds[i] = (i + 1) * size;
```

Comment:
This could possibly return an error because linear_update_bounds could be NULL. I'd recommend to check if linear_upper_bounds is NULL and do sth if yes

Reply:
Same in the other upper bounds functions.